### ISSUE TITLE
Shared object / DLL support

### ISSUE THREAD

**Issue by seadra** [tinygo-org/tinygo#254]
gc compiler has -buildmode=c-shared.
With LLVM, this should be possible, no?

**Comment by aykevl** [comment]
Yes, this is certainly possible. It's just that no one has asked for it yet.
It will need the extra `-buildmode` flag and some extra hooks, like an extra `-static` flag to the linker and  an extra hook to let the compiler emit PIC code instead of static code:

https://github.com/tinygo-org/tinygo/blob/5569cd1b6b3aff8f52052d5ff1028d9bb87fba24/compiler/compiler.go#L105

**Comment by JasonWeise** [comment]
This would be a great addition.
I am calling a Go library from Rust and would love to see it compiled with TinyGo instead.

Looking forward to this implementation :)

**Comment by sowelipililimute** [comment]
> Yes, this is certainly possible. It's just that no one has asked for it yet.
> It will need the extra `-buildmode` flag and some extra hooks, like an extra `-static` flag to the linker and  an extra hook to let the compiler emit PIC code instead of static code:
> 
> https://github.com/tinygo-org/tinygo/blob/5569cd1b6b3aff8f52052d5ff1028d9bb87fba24/compiler/compiler.go#L105

Is adding this really all that needs to be done? Seems like there would be more to do...

**Comment by aykevl** [comment]
It's likely some more plumbing needs to be done but that's the essence of it I think.

**Comment by ansiwen** [comment]
I'm using go with c-shared for building a small plugin .so, that basically only calls a rest API. So it mainly uses the http client and json parser. Because of the runtime, it's quite big. I think it would be very useful to compile this with tinygo instead.

**Comment by riaan-dau** [comment]
We have an existing C-codebase running bare-metal on 32-bit ARM and would like to make use of the `-buildmode=c-archive` flag to build a static Go library callable from C. Would this be possible if this feature is implemented?

**Comment by Jimmy2099** [comment]
It's almost 2023, how is going to support generate .dll or .lib

**Comment by wayneforrest** [comment]
I am also interested in this feature.

I am using a go .dll / .lib, called from C++; and the DLL is `13MB`, while the C++ code is only `1.5MB.`

**Comment by 31** [comment]
The Go `c-shared` support doesn't allow `dlclose` (or `FreeLibrary` on Windows), which makes it not ideal for plugins that need to be reloaded. I don't see any reason it'd be easier to support this with TinyGo than Go, but maybe this scenario has more value here? Go issue about this:

* https://github.com/golang/go/issues/11100

**Comment by turbolent** [comment]
Is there a simple way to do this manually at the moment? For example, compile to e.g. `.o` or `.ll`, and link the runtime?

**Comment by aykevl** [comment]
Not easily, you'd have to manually initialize the runtime, run package initializers, pass `-scheduler=none` because there is no scheduler, and make sure the functions aren't being called from more than one place.

Maybe I'll add support for this at some point since there seems to be demand for this feature.

**Comment by M09Ic** [comment]
Hi @aykevl, thanks so much for TinyGo — it's an amazing project and has been incredibly useful for our work!

We've been working on building a **Windows DLL** from TinyGo for FFI consumption by Rust/C programs ([chainreactors/rem](https://github.com/chainreactors/rem) — a network tunneling tool). After significant experimentation, we got a **working DLL solution** but hit a hard wall trying to produce a **static `.a` library** (c-archive equivalent). Here's what we found:

### What works: DLL via `ld.lld --dll`

```
tinygo build -o export.bc ./cmd/export_tinygo    # Go → LLVM bitcode
gcc -c dllmain.c -o dllmain.o                    # C bridge with dispatcher
ld.lld --dll -e DllMain ... export.bc + deps     # LTO link → DLL
```

We implemented a **thread dispatcher** pattern: \`main.main()\` runs a dispatch loop on the Go thread (never returns), and C bridge functions proxy FFI calls via shared globals + atomic spin-wait. This solves the "TinyGo runtime is thread-bound" problem. DLL works perfectly — verified with C (MinGW) and Rust (MSVC), including real network connections.

### What doesn't work: static `.a`

We tried the same bitcode merge + internalize approach that works for Rust ([llvm-link → opt -internalize → llc → llvm-ar](https://github.com/nicolo-ribaudo/tc39-proposal-structs)), but the resulting `.a` crashes at the first GC allocation when linked into an exe.

**Root cause we identified: bdwgc's `GC_register_data_segments()` / `GC_init_win32()`** calls `GetModuleHandle(NULL)` to find the module base, then scans the entire module's `.data`/`.bss` for GC roots. In DLL mode, this scans only the DLL's data — correct. In EXE mode (static `.a`), this scans the **entire EXE's data including the consumer's CRT**, causing GC corruption and ACCESS_VIOLATION on first `GC_malloc`.

### What we think would fix it

1. **`GC_set_no_dls(1)` + explicit `GC_add_roots()`** — tell bdwgc to only scan TinyGo's data segments, not the whole module
2. **Or compile bdwgc with `GC_DONT_REGISTER_MAIN_STATIC_DATA`** — skip automatic main module scanning
3. **Ideally: `-buildmode=c-archive` support** with proper runtime initialization separation (similar to Go's `gcc_libinit_windows.c` pattern)

### Our dispatcher pattern (available for reference)

The thread dispatcher we built could serve as a foundation for c-archive support:
- `main.main()` → `dispatchLoop()` (never returns, Go runtime stays alive)
- C bridge functions write to shared globals, signal via atomics, spin-wait for response
- All Go code executes on the Go thread (GC, scheduler, goroutines all work)
- Consumer thread never touches Go runtime directly

This pattern is verified working in DLL mode. The only blocker for `.a` is the bdwgc data segment scanning issue.

### Our setup
- TinyGo 0.40.1, LLVM 20, Windows x86_64
- MSYS2 MinGW for `ld.lld`, `llvm-link`, `opt`, `llc`
- Consumer: Rust (MSVC) and C (MinGW)

Would love to see c-archive/c-shared support in TinyGo. Happy to contribute or test if there's a path forward. Thanks again for all the work on TinyGo! 🙏

**Comment by hectorchu** [comment]
@M09Ic Have you tried not using the boehm gc (e.g. set gc to precise or conservative)?

**Comment by hectorchu** [comment]
Haha I got it working!

https://github.com/hectorchu/tinygo/commit/9450ee8181d2b316a1e836b8beea3b21b753b948

**Comment by hectorchu** [comment]
Works with this build command:

tinygo build -buildmode=c-archive -gc=conservative -scheduler=none

I've only made the necessary changes for darwin - you may need to modify windows or linux as appropriate.

**Comment by soypat** [comment]
So this is fixed? I think this message is relevant https://gophers.slack.com/archives/CDJD3SUP6/p1775818558605069

**Comment by hectorchu** [comment]
> So this is fixed? I think this message is relevant https://gophers.slack.com/archives/CDJD3SUP6/p1775818558605069

Well, what would you like to know? I have a thing or two I could say.

What I am now using, is this: #5334

But to answer your question: no, nothing has been fixed, nothing has been merged, and probably nothing ever will be.
